### PR TITLE
Transfer IAM Service Account

### PR DESCRIPTION
**Pull request by MikeMike41**

### CODE REVIEW

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 1]
```diff
@@ -2047,8 +2050,18 @@ private JsonObject populateMetaData(Response response) {
...
 			}
+
+			String readPolicy = new StringBuffer().append(TVaultConstants.SVC_ACC_POLICIES_PREFIXES.getKey(TVaultConstants.READ_POLICY)).append(IAMServiceAccountConstants.IAMSVCACC_POLICY_PREFIX).append(uniqueIAMSvcaccName).toString();
+			String writePolicy = new StringBuffer().append(TVaultConstants.SVC_ACC_POLICIES_PREFIXES.getKey(TVaultConstants.WRITE_POLICY)).append(IAMServiceAccountConstants.IAMSVCACC_POLICY_PREFIX).append(uniqueIAMSvcaccName).toString();
+			String denyPolicy = new StringBuffer().append(TVaultConstants.SVC_ACC_POLICIES_PREFIXES.getKey(TVaultConstants.DENY_POLICY)).append(IAMServiceAccountConstants.IAMSVCACC_POLICY_PREFIX).append(uniqueIAMSvcaccName).toString();
+
+			List<String> policiesToRemove = new ArrayList<>();
+			policiesToRemove.add(readPolicy);
+			policiesToRemove.add(writePolicy);
+			policiesToRemove.add(denyPolicy);
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:2061]: Not sure why we have to remove read and deny policies

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 2]
```diff
@@ -2156,9 +2159,14 @@ private JsonObject populateMetaData(Response response) {
...
 			policies.addAll(currentpolicies);
-			policies.remove(readPolicy);
-			policies.remove(writePolicy);
-			policies.remove(denyPolicy);
+			for (String policy : policiesToRemove) {
+				policies.remove(policy);
+				log.debug(JSONUtil.getJSON(ImmutableMap.<String, String>builder()
+						.put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER))
+						.put(LogMessage.ACTION, IAMServiceAccountConstants.REMOVE_USER_FROM_IAMSVCACC_MSG)
+						.put(LogMessage.MESSAGE, String.format("Removing policy from list - [%s]", policy))
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:2167]: Can we from which user and account? Will it provide additional information in logs?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 3]
```diff
@@ -5474,4 +5482,356 @@ private boolean updateIAMSvcAccSecretWithLatestExpiryDate(String token, String a
...
+	public ResponseEntity<String> transferIAMServiceAccountOwner(String token, UserDetails userDetails, IAMServiceAccountTransfer iamServiceAccountTransfer) throws IOException {
+		String iamSvcAccName = iamServiceAccountTransfer.getAwsAccountId() + "_" +
+				iamServiceAccountTransfer.getUserName();
+		String iamSvccAccPath = IAMServiceAccountConstants.IAM_SVCC_ACC_PATH + iamSvcAccName;
+		String iamSvccAccMetaPath = IAMServiceAccountConstants.IAM_SVCC_ACC_META_PATH + iamSvcAccName;
+		Response metaResponse = getMetadata(token, userDetails, iamSvccAccPath);
+		IAMServiceAccount iamSvcAcc = constructIAMSvcAccObjectFromMetadata(metaResponse);
+		IAMServiceAccount originalIAMSvcAcc = (IAMServiceAccount) SerializationUtils.clone(iamSvcAcc);
+
+		ResponseEntity<String> validationErrors = getTransferValidationErrors(token, iamSvcAcc, iamServiceAccountTransfer);
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5503]: Why is the method named as getTransferValidationErrors?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 4]
```diff
@@ -5474,4 +5482,356 @@ private boolean updateIAMSvcAccSecretWithLatestExpiryDate(String token, String a
...
+					.put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL)).build()));
+		} else {
+			log.error(JSONUtil.getJSON(ImmutableMap.<String, String>builder()
+					.put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER))
+					.put(LogMessage.ACTION, IAMServiceAccountConstants.IAM_SVCACC_TRANSFER_TITLE)
+					.put(LogMessage.MESSAGE,
+							String.format("Updating metadata for IAM Service Account [%s] failed.", iamSvccAccMetaPath))
+					.put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL)).build()));
+			return ResponseEntity.status(HttpStatus.MULTI_STATUS).body(
+					"{\"errors\":[\"Metadata update failed for IAM Service Account.\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5614]: Which account is it failed?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 5]
```diff
@@ -5474,4 +5482,356 @@ private boolean updateIAMSvcAccSecretWithLatestExpiryDate(String token, String a
...
+	 * Checks the user's authorization as well as validates
+	 * that the Service Account is not null and the new owner
+	 * is not the same as the existing owner.
+	 *
+	 * @param token
+	 * @param iamSvcAcc
+	 * @param iamSvcAccTransfer
+	 * @return ResponseEntity<String>
+	 */
+	private ResponseEntity<String> getTransferValidationErrors(String token, IAMServiceAccount iamSvcAcc,
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5677]: method naming convention?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 6]
```diff
@@ -5474,4 +5482,356 @@ private boolean updateIAMSvcAccSecretWithLatestExpiryDate(String token, String a
...
+															  IAMServiceAccountTransfer iamSvcAccTransfer) {
+		if (iamSvcAcc == null) {
+			log.error(JSONUtil.getJSON(ImmutableMap.<String, String>builder()
+					.put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER))
+					.put(LogMessage.ACTION, IAMServiceAccountConstants.IAM_SVCACC_TRANSFER_TITLE)
+					.put(LogMessage.MESSAGE, "Failed constructing metadata object.")
+					.put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL)).build()));
+			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+					"{\"errors\":[\"Failed to get metadata for this IAM Service Account.\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5686]: Is the message correct here?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java [hunk 7]
```diff
@@ -5474,4 +5482,356 @@ private boolean updateIAMSvcAccSecretWithLatestExpiryDate(String token, String a
...
+		}
+
+		if (iamSvcAccTransfer.getOwnerNtid().equals(iamSvcAcc.getOwnerNtid())) {
+			log.error(JSONUtil.getJSON(ImmutableMap.<String, String>builder()
+					.put(LogMessage.USER, ThreadLocalContext.getCurrentMap().get(LogMessage.USER))
+					.put(LogMessage.ACTION, IAMServiceAccountConstants.IAM_SVCACC_TRANSFER_TITLE)
+					.put(LogMessage.MESSAGE, "Failed to transfer IAM Service Account owner. The owner given is already the current owner.")
+					.put(LogMessage.APIURL, ThreadLocalContext.getCurrentMap().get(LogMessage.APIURL)).build()));
+			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+					"{\"errors\":[\"Failed to transfer IAM Service Account owner. The owner given is already the current owner.\"]}");
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5696]: Is the message appropriate here?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/IAMServiceAccountsController.java [hunk 8]
```diff
@@ -77,6 +65,20 @@
 		return iamServiceAccountsService.onboardIAMServiceAccount(token, iamServiceAccount, userDetails);
 	}
 
+	/**
+	 * Transfers an IAM service account from one owner to another
+	 * @param request
+	 * @param token
+	 * @return
+	 */
+	@ApiOperation(value = "${IAMServiceAccountsController.transferIAMServiceAccountOwner.value}", notes = "${IAMServiceAccountsController.transferIAMServiceAccountOwner.notes}")
```
Comment (sivanag1974) [inline tvaultapi/src/main/java/com/tmobile/cso/vault/api/v2/controller/IAMServiceAccountsController.java:74]: Is the message added to properties file?

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java (hunk 1 above)
Comment (MikeMike41) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:2061]: This was existing logic for removeUserFromIAMServiceAccount, I just refactored it out of the processAndRemoveUserPermissionFromIAMSvcAcc method so I could pass in the policy I wanted to remove from transfer method.

File: tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java (hunk 6 above)
Comment (MikeMike41) [reply tvaultapi/src/main/java/com/tmobile/cso/vault/api/service/IAMServiceAccountsService.java:5686]: This object should only ever be null if the metadata can't be retrieved and/or parsed, but I can change this verbiage if you think best.

**Pull request merged**